build-tools/src/main/java/org/elasticsearch/gradle/plugin/GenerateTestBuildInfoTask.java

```diff
@@ -0,0 +1,294 @@
+                }
+            }
+            if (path.length() > 18) {
+                path.append("/module-info.class");
+                JarEntry moduleEntry = jarFile.getJarEntry(path.toString());
+                if (moduleEntry != null) {
+                    try (InputStream inputStream = jarFile.getInputStream(moduleEntry)) {
+                        moduleName = extractModuleNameFromModuleInfo(inputStream);
+                    }
+                }
+            }
+        }
+
+        // if the jar is *not* multi-release then first look in
+        // module-info.class from the top-level if it exists
+        if (moduleName == null) {
+            JarEntry moduleEntry = jarFile.getJarEntry("module-info.class");
+            if (moduleEntry != null) {
+                try (InputStream inputStream = jarFile.getInputStream(moduleEntry)) {
+                    moduleName = extractModuleNameFromModuleInfo(inputStream);
+                }
+            }
+        }
+
+        // if the jar does *not* contain module-info.class
+        // check the manifest file for the module name
+        if (moduleName == null) {
+            JarEntry manifestEntry = jarFile.getJarEntry("META-INF/MANIFEST.MF");
+            if (manifestEntry != null) {
+                try (InputStream inputStream = jarFile.getInputStream(manifestEntry)) {
+                    Manifest manifest = new Manifest(inputStream);
+                    String amn = manifest.getMainAttributes().getValue("Automatic-Module-Name");
+                    if (amn != null) {
+                        moduleName = amn;
+                    }
+                }
+            }
+        }
+
+        // if the jar does not have module-info.class and no module name in the manifest
+        // default to the jar name without .jar and no versioning
+        if (moduleName == null) {
+            String jn = file.getName().substring(0, file.getName().length() - 4);
+            Matcher matcher = Pattern.compile("-(\\d+(\\.|$))").matcher(jn);
+            if (matcher.find()) {
+                jn = jn.substring(0, matcher.start());
+            }
+            jn = jn.replaceAll("[^A-Za-z0-9]", ".");
+            moduleName = jn;
+        }
+
+        return moduleName;
+    }
+
+    // find the first class and module when the class path entry is a directory
+    private void extractFromDirectory(File file, Map<String, String> classesToModules) throws IOException {
+        String className = extractClassNameFromDirectory(file);
+        String moduleName = extractModuleNameFromDirectory(file);
+
+        if (className != null && moduleName != null) {
+            classesToModules.put(className, moduleName);
+        }
+    }
+
+    // look through the directory to find the first unique class that isn't
+    // module-info.class (which may not be unique) and avoid anonymous classes
+    private String extractClassNameFromDirectory(File file) {
+        List<File> files = new ArrayList<>(List.of(file));
```

Comment:
Would it be simpler to just use `Files.walk` to traverse this directory?

Reply:
I have a commit that does this [here](https://github.com/prdoyle/elasticsearch/commit/33ce238c406e0e91b85a7afec214fbb43442e640). I can push it to Jack's repo if we like it.

(Fun fact: this uses a non-denotable type!)